### DISCUSSION TITLE
Unable to override `transformIgnorePatterns` in `next/jest`

### DISCUSSION THREAD

**Discussion by czebe** [vercel/next.js#34774] [Ideas]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:54 PST 2021; root:xnu-8019.61.5~1/RELEASE_X86_64
    Binaries:
      Node: 16.14.0
      npm: 8.3.1
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.1.1-canary.2
      react: 17.0.2
      react-dom: 17.0.2

### Describe the Bug

Some libraries ship `es6` code and require to use `next-transpile-modules`.

It could be easy to transpile these by specifiying in `transformIgnorePatterns`, but the default config is not allowing overrides.

https://github.com/vercel/next.js/blob/cc5345bba852077c332ddc0b75279a0f96d4321a/packages/next/build/jest/jest.ts#L118

### Expected Behavior

Fully replace `transformIgnorePatterns` if it's specified in `resolvedJestConfig`.

### To Reproduce

N/A

**Comment by ijjk** [comment]
Hi, I'm not sure we should completely throw away the default `transformIgnorePatterns` if one is provided since this could accidentally cause `.next` and all of `node_modules` to be transformed. Instead you can augment the built config e.g.

```js
const nextJest = require('next/jest')

const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: './',
})

// Add any custom config to be passed to Jest
const customJestConfig = {
  // Add more setup options before each test is run
  // setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
  // if using TypeScript with a baseUrl set to the root directory then you need the below for alias' to work
  moduleDirectories: ['node_modules', '<rootDir>/'],
  testEnvironment: 'jest-environment-jsdom',
}

module.exports = async (...args) => {
  const fn = createJestConfig(customJestConfig)
  const res = await fn(...args)
  
  res.transformIgnorePatterns = res.transformIgnorePatterns.map(pattern => {
    if (pattern === '/node_modules/') {
      return '/node_modules(?!/excluded-module)/'
    }
    return pattern
  })
  
  return res
}
```

**Reply by davidcalhoun** [reply]
Thanks for the code sample, it's running great here!

Note that right now it's unexpected that some options in `customJestConfig` work (like `setupFilesAfterEnv` and `testEnvironment`) while others don't (`transformIgnorePatterns`).  Probably makes sense to add this caveat to these docs: https://nextjs.org/docs/testing#setting-up-jest-with-the-rust-compiler

**Reply by wadehammes** [reply]
Typescript is not a big fan of the (...args), do you know what they should be typed for this?

**Reply by ChrisVilches** [reply]
This is the only workaround that worked for me to convert a `.js` file in `node_modules` with non-standard syntax.
